tests/utils_tests/test_choices.py

```diff
@@ -0,0 +1,216 @@
+        # Canonical form is list of 2-tuple, but ensure lists all the way down work.
+        choices = [
+            ["C", _("Club")],
+            ["D", _("Diamond")],
+            ["H", _("Heart")],
+            ["S", _("Spade")],
+        ]
+        self.assertEqual(normalize_field_choices(choices), self.expected)
+
+    def test_iterator_non_canonical(self):
+        # Canonical form is list of 2-tuple, but ensure lists all the way down work.
+        def generator():
+            yield ["C", _("Club")]
+            yield ["D", _("Diamond")]
+            yield ["H", _("Heart")]
+            yield ["S", _("Spade")]
+
+        choices = generator()
+        self.assertEqual(normalize_field_choices(choices), self.expected)
+
+    def test_nested_iterable_non_canonical(self):
+        # Canonical form is list of 2-tuple, but ensure lists all the way down work.
+        choices = [
+            ["Audio", [["vinyl", _("Vinyl")], ["cd", _("CD")]]],
+            ["Video", [["vhs", _("VHS Tape")], ["dvd", _("DVD")]]],
+            ["unknown", _("Unknown")],
+        ]
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+
+    def test_nested_iterator_non_canonical(self):
+        # Canonical form is list of 2-tuple, but ensure lists all the way down work.
+        def generator():
+            yield ["Audio", [["vinyl", _("Vinyl")], ["cd", _("CD")]]]
+            yield ["Video", [["vhs", _("VHS Tape")], ["dvd", _("DVD")]]]
+            yield ["unknown", _("Unknown")]
+
+        choices = generator()
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+
+    def test_nested_mixed_mapping_and_iterable(self):
+        # Although not documented as it's better to stick to either mappings or
+        # iterables, ensure that this works as expected.
+        choices = {
+            "Audio": [("vinyl", _("Vinyl")), ("cd", _("CD"))],
+            "Video": [("vhs", _("VHS Tape")), ("dvd", _("DVD"))],
+            "unknown": _("Unknown"),
+        }
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+        choices = [
+            ("Audio", {"vinyl": _("Vinyl"), "cd": _("CD")}),
+            ("Video", {"vhs": _("VHS Tape"), "dvd": _("DVD")}),
+            ("unknown", _("Unknown")),
+        ]
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+
+    def test_iterable_set(self):
+        # Although not documented as sets are unordered which results in randomised
+        # order in form fields, ensure that this works as expected. Note that consistent
+        # ordering in migrations is enforced by the migrations serializer.
+        choices = {
+            ("C", _("Club")),
+            ("D", _("Diamond")),
+            ("H", _("Heart")),
+            ("S", _("Spade")),
+        }
+        self.assertEqual(
+            sorted(normalize_field_choices(choices)),
+            sorted(self.expected),
+        )
```

Comment:
Yeah, so this is another thing that will have occurred in the wild. Migrations handle stability by sorting sets which isn't ideal. In forms, the order will be arbitrary. I had toyed with adding a case to the match statement to raise a warning about this à la the warning about unordered results in pagination. Something we could consider?